### ISSUE TITLE
SSL/TLS checks plugin

### ISSUE THREAD

**Issue by treadie** [Arachni/arachni#452]
Would be cool to have a plugin that did some comprehensive SSL/TLS checks, similar to what SSL LABS server test does.

Look here for some examples: https://www.ssllabs.com/ssltest/

**Comment by Zapotek** [comment]
Not sure that the HTTP library I'm using supports this, see: typhoeus/typhoeus/issues/427

We'll see how it goes.

**Comment by dmitris** [comment]
you may be also interested in (or even get involved in) the https://github.com/drwetter/testssl.sh project (though SSL testing is done in a different way)

**Comment by bararchy** [comment]
@Zapotek  Well, if you want to are more then welcome to use all\parts of my project (all in pure ruby) 

https://github.com/bararchy/ruby-SSLscanner

**Comment by Zapotek** [comment]
@bararchy Problem is I need to do this via Arachni's global HTTP interface in order to enforce a bunch of user options and restrictions -- the interface is based on Typhoeus, which in turn delegates to `libcurl`.

However, I did find out that the `libcurl` debugging data I use to extract raw HTTP messages for the reports also include the SSL handshake traffic.
Problem is, I'm not sure how to feed that traffic to the SSL libs in order to get the context and other data in a useful way.
I could create 2 local sockets, feed them the raw captured data and then use their `SSLSocket` wrappers to extract the context, or something equally cumbersome, but I'd rather it doesn't come to this.

You seem much more comfortable with the SSL API than me, is there any way to do this painlessly?

**Comment by bararchy** [comment]
@Zapotek Well, feeding raw TCP\SSL data to a SSLSocket sounds like a very hard thing to achieve in Ruby, I would not go there.

What I would do, is maybe create a external "SSL Cipher Check" plugin, and call it using 'require' or just using system('sslscan.rb') and parse the data.
Would something like that be acceptable with the way you work ? 
Maybe even thinking about a way to output XML or HTML data from the plugin so the parsing will be better handled.

**Comment by Zapotek** [comment]
I'm already doing MiTM stuff in order to intercept SSL browser connections, and I've had to rewrite that code in a lower level fashion for the new and better HTTP proxy that will be merged 1-2 releases down the road, so I've already gone through that ordeal once or twice.

To be frank I'd rather do some local MiTM socket hackery than rely on external applications.
By that I don't mean 3rd party of course, I just don't like having to manage other executables.

And there are the user specified options and restrictions that I need to respect (timeouts, proxies, etc.), which I won't be able to control all that well if I delegate to another program.

Until I find a solution I like, you can run external applications already and capture their output to be included in the reports via the `exec` plugin:

```
$ ./bin/arachni --plugins-list=exec
Arachni - Web Application Security Scanner Framework v2.0dev
   Author: Tasos "Zapotek" Laskos <tasos.laskos@arachni-scanner.com>

           (With the support of the community and the Arachni Team.)

   Website:       http://arachni-scanner.com
   Documentation: http://arachni-scanner.com/wiki




 [~] Available plugins:

 [*] exec:
--------------------
Name:           Exec
Description:
Calls external executables at different scan stages.

The following variables can be used in the executable string:

* `__URL__`: Target URL.
* `__URL_SCHEME__`: URL scheme (http/https).
* `__URL_HOST__`: URL host.
* `__URL_PORT__`: URL port.
* `__STAGE__`: Scan stage (pre/during/post).
* `__ISSUE_COUNT__`: Amount of logged issues.
* `__SITEMAP_SIZE__`: Amount of discovered pages.
* `__FRAMEWORK_RUNTIME__`: Scan runtime.
* `__FRAMEWORK_STATUS__`: Framework status.

**Example:**

The following:

    echo "__URL__ __URL_SCHEME__ __URL_HOST__ __URL_PORT__ __STAGE__ __ISSUE_COUNT__ __SITEMAP_SIZE__ __FRAMEWORK_RUNTIME__ __FRAMEWORK_STATUS__"

Will result in:

    http://testfire.net/ http testfire.net 80 post 0 2 3.906482517 cleanup

_Will not work over RPC._

Options:
 [~]    pre - Executable to be called prior to the scan.
 [~]    Type:        string
 [~]    Default:     
 [~]    Required?:   false

 [~]    during - Executable to be called in parallel to the scan.
 [~]    Type:        string
 [~]    Default:     
 [~]    Required?:   false

 [~]    post - Executable to be called after the scan.
 [~]    Type:        string
 [~]    Default:     
 [~]    Required?:   false

Author:         Tasos "Zapotek" Laskos <tasos.laskos@arachni-scanner.com>
Version:        0.1
Path:   /home/zapotek/workspace/arachni/components/plugins/exec.rb
```

**Comment by bararchy** [comment]
@Zapotek The exec plugin looks really cool, I'll use it :) 

About the sockets, the issue with scanning SSL Ciphers\Protocols is that you need to work by way of elimination, for example

Client = SSL3, TLS1.0, TLS1.1, TLS1.2 :: RC4

Server = SSL3, TLS1.2 :: RC4

Client =SSL3 :: RC4 => Server? OK
Client =TLS1.0 :: RC4 => Server? Not Supported
Client =TLS1.1 :: RC4 => Server? Not Supported
Client =TLS1.2 :: RC4 => Server? OK

So basically we will connect using each protocol with each available cipher once to see  if the server supports it.
This in turn means you need to define a new SSLSocket each time and enable only one Protocol\Cipher pair and try to connect to the remote server.

**Comment by Zapotek** [comment]
That's the easy part, if I can just go over a list of options and determine if they're supported just by the success of the HTTP request then I don't have to worry about anything else.

And then I'd use my socket hackery to extract keys and certificates and whatnot from the traffic associated with those requests -- if at all possible of course.
